File: uvaonlinejudge/src/volume_ci/P10194_fernandohbc.java
/*
 * MyPCS - My Programming Contests Solutions
 * 
 * Copyright (C) 2007 - 2011  Fernando Cardoso
 * email: fernandohbc@gmail.com / twitter: fernando_hbc
 * 
 * The source code of this program is free software; you can do whatever you
 * want with it since you always keep the name and contact of the original
 * author. Also, if you redistribute a modified version, make sure you explicit
 * the portions of the code that have been modified
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. In other words, use it
 * at your own risk.
 * 
 * If you find this a little useful, you can buy the author some beer.
 * If you find this really useful, you can buy a kid a book, and help making
 * the world more literate.
 */
package volume_ci;

import java.io.OutputStreamWriter;
import java.io.PrintWriter;
import java.io.UnsupportedEncodingException;
import java.util.Arrays;
import java.util.Scanner;

/**
 * 10194 - Football (aka Soccer)
 * 
 * @author Fernando Cardoso (fernandohbc@gmail.com / Twitter: fernando_hbc)
 * 
 */
public class P10194_fernandohbc {
    public static void main(String[] args) throws UnsupportedEncodingException {
        Scanner cin = new Scanner(System.in, "ISO-8859-1");
        PrintWriter cout = new PrintWriter(new OutputStreamWriter(System.out,
                "ISO-8859-1"));

        int nOfTournment = Integer.parseInt(cin.nextLine());
        for (int t = 0; t < nOfTournment; t++) {
            if (t != 0) {
                cout.write("\n");
            }
            // Reads the tournment name
            String tournmentName = cin.nextLine();

            // How many teams?
            int nOfTeams = Integer.parseInt(cin.nextLine());
            Team[] teams = new Team[nOfTeams];
            for (int team = 0; team < nOfTeams; team++) {
                // Reads the name of the team
                teams[team] = new Team(cin.nextLine());
            }

            // How many games?
            int nOfGames = Integer.parseInt(cin.nextLine());
            for (int game = 0; game < nOfGames; game++) {
                // Reads the game
                parseGame(cin.nextLine(), teams);
            }

            // Sorts
            Arrays.sort(teams);

            // Prints
            cout.write(tournmentName + "\n");
            cout.flush();
            for (int team = 0; team < nOfTeams; team++) {
                Team tm = teams[team];
                System.out.printf("%d) %s %dp, %dg (%d-%d-%d), %dgd (%d-%d)\n",
                        team + 1, tm.getName(), tm.getPoints(),
                        tm.getGamesPlayed(), tm.getWins(), (tm.getGamesPlayed()
                                - tm.getWins() - tm.getLosses()),
                        tm.getLosses(),
                        (tm.getGoalsScored() - tm.getGoalsAgainst()),
                        tm.getGoalsScored(), tm.getGoalsAgainst());
            }
        }

        cout.flush();
    }

    private static void parseGame(String line, Team[] teams) {
        // First team
        String tok = line.substring(0, line.indexOf("#"));
        String rest = line.substring(line.indexOf("#") + 1);
        Team firstTeam = locateTeam(teams, tok);

        // First team score
        tok = rest.substring(0, rest.indexOf("@"));
        rest = rest.substring(rest.indexOf("@") + 1);
        int firstTeamScore = Integer.parseInt(tok);

        // Second team score
        tok = rest.substring(0, rest.indexOf("#"));
        rest = rest.substring(rest.indexOf("#") + 1);
        int secondTeamSocre = Integer.parseInt(tok);

        // Second team
        Team secondTeam = locateTeam(teams, rest);

        Team winner = null;
        Team loser = null;
        if (firstTeamScore > secondTeamSocre) {
            winner = firstTeam;
            loser = secondTeam;
        } else if (secondTeamSocre > firstTeamScore) {
            winner = secondTeam;
            loser = firstTeam;
        }

        // Has a winner?
        if (winner != null) {
            winner.addPoints(3);
            winner.addWin();

            loser.addLoss();
        } else {
            firstTeam.addPoints(1);
            secondTeam.addPoints(1);
        }

        // Update scores
        firstTeam.addGoalsScored(firstTeamScore);
        firstTeam.addGoalsAgainst(secondTeamSocre);
        secondTeam.addGoalsAgainst(firstTeamScore);
        secondTeam.addGoalsScored(secondTeamSocre);

        // Games played
        firstTeam.addGamePlayed();
        secondTeam.addGamePlayed();
    }

    private static Team locateTeam(Team[] teams, String tok) {
        for (Team team : teams) {
            if (team.getName().equals(tok)) {
                return team;
            }
        }
        return null;
    }
}

class Team implements Comparable<Team> {

    private int points = 0;
    private int wins = 0;
    private int goalsScored = 0;
    private int goalsAgainst = 0;
    private int losses = 0;
    private int gamesPlayed = 0;
    private String name;

    public Team(String name) {
        this.name = name;
    }

    public int getGoalsAgainst() {
        return this.goalsAgainst;
    }

    public int getGoalsScored() {
        return this.goalsScored;
    }

    public int getLosses() {
        return this.losses;
    }

    public int getWins() {
        return this.wins;
    }

    public int getGamesPlayed() {
        return this.gamesPlayed;
    }

    public int getPoints() {
        return this.points;
    }

    public void addGamePlayed() {
        this.gamesPlayed++;
    }

    public void addGoalsAgainst(int goals) {
        this.goalsAgainst += goals;
    }

    public void addGoalsScored(int goals) {
        this.goalsScored += goals;
    }

    public void addLoss() {
        this.losses++;
    }

    public void addWin() {
        this.wins++;
    }

    public void addPoints(int i) {
        this.points += i;
    }

    public String getName() {
        return this.name;
    }

    @Override
    public int compareTo(Team other) {
        int diff = other.points - this.points;
        if (diff != 0) {
            return diff;
        }

        diff = other.wins - this.wins;
        if (diff != 0) {
            return diff;
        }

        int aGd = this.goalsScored - this.goalsAgainst;
        int bGd = other.goalsScored - other.goalsAgainst;
        diff = bGd - aGd;
        if (diff != 0) {
            return diff;
        }

        diff = other.goalsScored - this.goalsScored;
        if (diff != 0) {
            return diff;
        }

        diff = this.gamesPlayed - other.gamesPlayed;
        if (diff != 0) {
            return diff;
        }

        return this.name.toUpperCase().compareTo(other.name.toUpperCase());
    }
}

File: uvaonlinejudge/src/volume_ci/P10194_fernandohbc.cpp
/*******************************************************************************
 * MyPCS - My Programming Contests Solutions
 * 
 * Copyright (C) 2007 - 2011  Fernando Cardoso
 * email: fernandohbc@gmail.com / twitter: fernando_hbc
 * 
 * The source code of this program is free software; you can do whatever you
 * want with it since you always keep the name and contact of the original
 * author. Also, if you redistribute a modified version, make sure you explicit
 * the portions of the code that have been modified
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. In other words, use it
 * at your own risk.
 * 
 * If you find this a little useful, you can buy the author some beer.
 * If you find this really useful, you can buy a kid a book, and help making
 * the world more literate.
 ******************************************************************************/
#include <iostream>
#include <algorithm>

#define TEAMNAME 31
#define MAXLINE 70

/*
 * TODO: This is a WA code as of 03 March 2011
 * Must be corrected to be accepted.
 */

using namespace std;

struct Team {
  int points;
  int wins;
  int goalsScored;
  int goalsAgainst;
  int losses;
  int gamesPlayed;
  char name[TEAMNAME];
};

Team * locateTeam( Team * teams, char * teamName) {
  int i = 0;
  while (strcmp(teams[i].name, teamName) != 0 ) {
    i++;
  }
  return teams + i;
}

void parseGame(char * encodedGame, Team teams[]) {
  // First team
  char * tok = strtok(encodedGame, "#");
  Team * firstTeam = locateTeam(teams, tok);

  // First team score
  tok = strtok(NULL, "@");
  int firstTeamScore = atoi(tok);

  // Second team score
  tok = strtok(NULL, "#");
  int secondTeamSocre = atoi(tok);

  //Second team
  tok = strtok(NULL, "#");
  Team * secondTeam = locateTeam(teams, tok);

  Team * winner = NULL;
  Team * loser = NUL:
  if ( firstTeamScore > secondTeamSocre ) {
    winner = firstTeam;
    loser = secondTeam;
  } else if ( secondTeamSocre > firstTeamScore ){
    winner = secondTeam;
    loser = firstTeam;
  }

  // Has a winner?
  if (winner) {
    winner->points += 3;
    winner->wins += 1;

    loser->losses += 1;
  } else {
    firstTeam->points += 1;
    secondTeam->points += 1;
  }


  // Update scores
  firstTeam->goalsScored += firstTeamScore;
  firstTeam->goalsAgainst += secondTeamSocre;
  secondTeam->goalsAgainst += firstTeamScore;
  secondTeam->goalsScored += secondTeamSocre;

  // Games played
  firstTeam->gamesPlayed++;
  secondTeam->gamesPlayed++;
}

void initializeTeam(Team& team, char * teamName) {
  strcpy(team.name, teamName);
  team.points = 0;
  team.wins = 0;
  team.goalsScored = 0;
  team.goalsAgainst = 0;
  team.losses = 0;
  team.gamesPlayed = 0;
}

bool compareTeams(Team a, Team b) {
  if ( a.points > b.points ) {
    return true;
  } else if ( a.points < b.points ) {
    return false;
  }

  if ( a.wins > b.wins ) {
    return true;
  } else if ( a.wins < b.wins ) {
    return false;
  }

  int aGd = a.goalsScored - a.goalsAgainst;
  int bGd = b.goalsScored - b.goalsAgainst;
  if ( aGd > bGd ) {
    return true;
  } else if ( aGd < bGd ) {
    return false;
  }

  if ( a.goalsScored > b.goalsScored ) {
    return true;
  } else if ( a.goalsScored < b.goalsScored ){
    return false;
  }

  if ( a.gamesPlayed < b.gamesPlayed ) {
    return true;
  } else if ( a.gamesPlayed > b.gamesPlayed ) {
    return false;
  }

  return strcasecmp(a.name, b.name) < 0;
}

int main() {
  char line[MAXLINE];

  cin.getline(line, MAXLINE);
  int nOfTournment = atoi(line);
  for (int t = 0; t < nOfTournment; t++) {
    if ( t != 0 ) {
      cout << endl;
    }
    // Reads the tournment name
    cin.getline(line, MAXLINE);
    int lengthTournmentName = strlen(line);
    char tournamentName[lengthTournmentName + 1];
    strcpy(tournamentName, line);

    // How many teams?
    cin.getline(line, MAXLINE);
    int nOfTeams = atoi(line);
    Team teams[nOfTeams];
    for ( int team = 0; team < nOfTeams; team++) {
      // Reads the name of the team
      cin.getline(line, MAXLINE);
      initializeTeam(teams[team], line);
    }

    // How many games?
    cin.getline(line, MAXLINE);
    int nOfGames = atoi(line);
    for ( int game = 0; game < nOfGames; game++ ) {
      // Reads the game
      cin.getline(line, MAXLINE);
      parseGame(line, teams);
    }

    // Sorts
    sort(teams, teams + nOfTeams, compareTeams);

    // Prints
    cout << tournamentName << endl;
    for ( int team = 0; team < nOfTeams; team ++ ) {
      Team tm = teams[team];
      printf("%d) %s %dp, %dg (%d-%d-%d), %dgd (%d-%d)\n", team+1, tm.name,
          tm.points, tm.gamesPlayed, tm.wins,
          (tm.gamesPlayed-tm.wins-tm.losses), tm.losses,
          (tm.goalsScored - tm.goalsAgainst), tm.goalsScored,
          tm.goalsAgainst);
    }
  }
  return 0;
}
